### DISCUSSION TITLE
Help to identify origin of (visible) increase in DNS Server's RAM usage from v12 to v13

### DISCUSSION THREAD

**Discussion by shalafi99** [TechnitiumSoftware/DnsServer#1111] [Q&A] [answered]
Hello @ShreyasZare,

I have been running Technitium DNS Server in 2 cloud virtual machines for 6 months or so.

They have just 1 GB RAM, but with the Windows Server core edition it has been quite manageable (probably because I host only one  zone, also use them as caching resolvers for public names, QPM rate limiting has been set to inhibit excessive usage by others).
As it turns out, I only have to increase the machine's RAM (temporarily) to install OS updates.

Since the very beginning, I have had a 2nd DNS Server instance inside each machine to function as a root zone secondary for the 1st DNS Server instance. (Like your [blog ](https://blog.technitium.com/2021/07/running-root-server-locally-on-your-dns.html)describes)

This is the usual memory footprint for both DNS services:
```
Name         Id Memory (MB)
----         -- -----------
DnsService 1532       73.49
DnsService 1524        23.2
```

The addition of ZONEMD you announced a few weeks back (with version 13) caught my interest - I had an expectation of being able to lower the memory footprint (even if just a little) by moving the secondary root zone into the primary DNS Server service (thus eliminating the overhead of a 2nd running windows service).

Today I upgraded 12.2.1 to 13.1.1 in one of the servers. All I did was:
1) stopped both DNS services
2) upgraded DNS Server
3) started **only** the primary DNS service
4) deleted the forwarder zone for the root zone
5) added the secondary root zone (just one click, like you announced)
(The server was not restarted)

But the hoped for expectation was actually something of a surprise:
```
Name        Id Memory (MB)
----        -- -----------
DnsService 152      157.46
```

A single 13.1.1 instance is consuming 50% more RAM than the two 12.2.1 instances I had before.
Any idea which new/updated features to 13.1.1 from 12.2.1 could be responsible for this additional RAM usage?

The DNS Apps installed are:
Drop Requests
Failover
Query Logs (Sqlite)

The only optional DNS Server protocol configured is DoT.
Blocking is configured, with one of StevenBlack's preset host block lists
Max cache size is 500.

Regards

**Comment by ShreyasZare** [comment]
Thanks for asking. The memory is managed by the garbage collector in the .NET runtime that the DNS server is running on. When the DNS server starts, there are many processes that run like loading the block lists which consume memory. The memory you see used will get freed automatically after a while as the garbage collector runs and de-allocate things that are not in use anymore. 

Another thing, the max cache size you have is very low and using the default value of 10,000 is recommended. The records do not take that much memory and the server can easily hold even more than 10k records on a server with 1GB RAM. Low max cache size will cause the DNS server to remove usable records and will cause it to fetch them again and again when it receives requests, decreasing the overall performance.

**Comment by shalafi99** [comment]
Thanks for the explanation, @ShreyasZare.
The server has a restart schedule and I missed looking at the RAM consumption before that reboot.

For the moment, what I see is this:

```
PS C:\> (Get-Date) - (Get-CimInstance -ClassName Win32_OperatingSystem).LastBootUpTime | ForEach-Object { "{0} days, {1} hours, {2} minutes, {3} seconds" -f $_.Days, $_.Hours, $_.Minutes, $_.Seconds }
1 days, 17 hours, 2 minutes, 59 seconds
PS C:\>
PS C:\> Get-Process | Where-Object { $_.Name -like "*dns*" } |
>> Sort-Object WorkingSet -Descending |
>> Select-Object -First 10 Name, ID, @{Name="Memory (MB)";Expression={[math]::Round($_.WorkingSet / 1MB, 2)}}

Name         Id Memory (MB)
----         -- -----------
DnsService 1700      110.25
```

That is still a bit higher (~15 MB) than the sum for the 2 instances with the older version.
(I will check again in a day or two)

**Reply by ShreyasZare** [reply]
Restart will show more RAM being used since the DNS server will be loading the block lists and other things which will use some memory. It will take some time for the memory usage to go down.

**Reply by shalafi99** [reply]
New check, just a bit short of 40h later:

```
PS C:\> (Get-Date) - (Get-CimInstance -ClassName Win32_OperatingSystem).LastBootUpTime | ForEach-Object { "{0} days, {1} hours, {2} minutes, {3} seconds" -f $_.Days, $_.Hours, $_.Minutes, $_.Seconds }
3 days, 7 hours, 58 minutes, 9 seconds
PS C:\>
PS C:\> Get-Process | Where-Object { $_.Name -like "*dns*" } |
>> Sort-Object WorkingSet -Descending |
>> Select-Object -First 10 Name, ID, @{Name="Memory (MB)";Expression={[math]::Round($_.WorkingSet / 1MB, 2)}}

Name         Id Memory (MB)
----         -- -----------
DnsService 1700       117.2
```

Given no changes in configuration or usage happen, I suppose the RAM usage for this setup I got will not ever dip below 100 MB... right? (_Usage patterns for this DNS service have been much predictable over these past months._)

Thanks

**Reply by ShreyasZare** [reply, marked as answer]
For programs developed in C/C++, for example, can precisely answer such questions. Memory usage pattern is not predictable with runtimes that use garbage collection.
